awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx

```diff
@@ -1,94 +1,688 @@
+      }
+    }
+
+    // Add the new links
+    parents.forEach(parentId => {
+      children.forEach(child => {
+        if (parentId === 1) {
+          // We only want to create a link from the start node to this node if it
+          // doesn't have any other parents
+          if (linkParentMapping[child.id].length === 1) {
+            newLinks.push({
+              source: { id: parentId },
+              target: { id: child.id },
+              linkType: 'always',
+              type: 'link',
+            });
+          }
+        } else if (!linkParentMapping[child.id].includes(parentId)) {
+          newLinks.push({
+            source: { id: parentId },
+            target: { id: child.id },
+            linkType: child.linkType,
+            type: 'link',
+          });
+        }
+      });
+    });
+    // need to track that this node has been deleted if it's not new
+
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setNodeToDelete(null);
+    setNodes(newNodes);
+    setLinks(newLinks);
+  };
+
+  const updateLink = linkType => {
+    const newLinks = [...links];
+    newLinks.forEach(link => {
+      if (
+        link.source.id === linkToEdit.source.id &&
+        link.target.id === linkToEdit.target.id
+      ) {
+        link.linkType = linkType;
+      }
+    });
+
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setLinkToEdit(null);
+    setLinks(newLinks);
+  };
+
+  const startDeleteLink = link => {
+    const parentMap = {};
+    links.forEach(existingLink => {
+      if (!parentMap[existingLink.target.id]) {
+        parentMap[existingLink.target.id] = [];
+      }
+      parentMap[existingLink.target.id].push(existingLink.source.id);
+    });
+
+    link.isConvergenceLink = parentMap[link.target.id].length > 1;
+
+    setLinkToDelete(link);
+  };
+
+  const deleteLink = () => {
+    const newLinks = [...links];
+
+    for (let i = newLinks.length; i--; ) {
+      const link = newLinks[i];
+
+      if (
+        link.source.id === linkToDelete.source.id &&
+        link.target.id === linkToDelete.target.id
+      ) {
+        newLinks.splice(i, 1);
+      }
+    }
+
+    if (!linkToDelete.isConvergenceLink) {
+      // Add a new link from the start node to the orphaned node
+      newLinks.push({
+        source: {
+          id: 1,
+        },
+        target: {
+          id: linkToDelete.target.id,
+        },
+        linkType: 'always',
+        type: 'link',
+      });
+    }
+
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setLinkToDelete(null);
+    setLinks(newLinks);
+  };
+
+  const selectSourceNodeForLinking = sourceNode => {
+    const newNodes = [...nodes];
+    const parentMap = {};
+    const invalidLinkTargetIds = [];
+    // Find and mark any ancestors as disabled to prevent cycles
+    links.forEach(link => {
+      // id=1 is our artificial root node so we don't care about that
+      if (link.source.id !== 1) {
```

Comment:
Consider checking for the opposite and returning early:

```
if (link.source.id === 1) {
  return;
}
```

This prevents nested `if`s and simplifies the logic a little

Reply:
👍 on it